src/pages/home/sidebar/SidebarLinks.js

```diff
@@ -104,58 +123,63 @@ class SidebarLinks extends React.Component {
-            const hasDraftCommentChanged = previousDraftComments[report] !== nextDraftComments[report];
-            const isNotANewDraftComment = !reportsWithNewDraftComments.includes(report);
-            const isNotARemovedDraftComment = !reportsWithRemovedDraftComments.includes(report);
-
-            return didDraftCommentExistPreviously && hasDraftCommentChanged && isNotANewDraftComment && isNotARemovedDraftComment;
-        });
-
-        const allReportsWithDraftCommentChanges = [
-            ...reportsWithNewDraftComments,
-            ...reportsWithRemovedDraftComments,
-            ...reportsWithEditedDraftComments,
-        ];
-
-        const activeReportID = this.props.currentlyViewedReportID;
-        const reportKey = `${ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT}${activeReportID}`;
-
-        // Do not re-order reports if draft comment changes are only in the current report.
-        if (allReportsWithDraftCommentChanges.length === 1 && allReportsWithDraftCommentChanges.includes(reportKey)) {
-            return false;
+        // If any reports have new unread messages, re-order the list
+        const nextUnreadReports = SidebarLinks.getUnreadReports(nextProps.reports);
+        const hasNewUnreadReports = nextUnreadReports.length > 0
+            && _.some(nextUnreadReports,
+                nextUnreadReport => !_.some(unreadReports, unreadReport => unreadReport.reportID === nextUnreadReport.reportID));
+        if (hasNewUnreadReports) {
+            return true;
+        }
+
+        // Do not re-order if the active report has a draft and vice versa.
+        if (nextProps.currentlyViewedReportID) {
+            const hasActiveReportDraft = lodashGet(nextProps.reportsWithDraft, `${ONYXKEYS.COLLECTION.REPORTS_WITH_DRAFT}${nextProps.currentlyViewedReportID}`, false);
+            return !hasActiveReportDraft;
         }
+
         return true;
     }
 
+    constructor(props) {
+        super(props);
+        this.state = {
+            currentlyViewedReportID: props.currentlyViewedReportID,
+            orderedReports: [],
+            unreadReports: SidebarLinks.getUnreadReports(props.reports),
+        };
+    }
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+        const shouldReorder = SidebarLinks.shouldReorder(nextProps, prevState.orderedReports, prevState.currentlyViewedReportID, prevState.unreadReports);
+        const recentReports = SidebarLinks.getRecentReports(nextProps);
+        const orderedReports = shouldReorder
+            ? recentReports
+            : _.map(prevState.orderedReports,
+                orderedReport => _.chain(recentReports)
+                    .filter(recentReport => orderedReport.reportID === recentReport.reportID)
+                    .first()
+                    .value());
+
+        return {
+            orderedReports,
+            currentlyViewedReportID: nextProps.currentlyViewedReportID,
+            unreadReports: SidebarLinks.getUnreadReports(nextProps.reports),
```

Comment:
I assume same here if we making these changes.
```suggestion
            unreadReports: SidebarLinks.getUnreadReports(nextProps.reports || {}),
```